### PR TITLE
Fix immersive mode panic on three.js rollercoaster on hololens

### LINKED ISSUE #20595: Implement component check in Image2D methods
I'm not sure we implement the following check:

> `GL_INVALID_OPERATION` is generated if the currently bound framebuffer's format does not contain a superset of the components required by the base *format* of *internalformat*.

https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.8
https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glCopyTexImage2D.xml
https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glCopyTexSubImage2D.xml

### LINKED ISSUE #24083: XR framebuffer creation panics when setting up texture
When running the three.js webvr rollercoaster example in immersive, I get GL errors when we try to bind the texture that we've just created as part of creating an XRWebGLLayer. My theory is that the webgl commands end up being interleaved with other GL processing from webrender in such a way that the state we rely upon is messed up. I have a patch to move these operations to run in the webgl thread in an atomic unit as part of a single WebGLCommand, and this makes the GL errors go away.

### PR DESCRIPTION
**Pull request by jdm**
We have some special logic about texture formats when creating drawing buffers for WebGL that needs to be shared with the code that creates a separate framebuffer for immersive mode.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #24083 and fix #20595.
- [x] These changes do not require tests because no CI for hololens; tested manually in the emulator.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/24242)

### CODE REVIEW

**Review by asajeffrey — APPROVED** [review]
I'm a bit uncomfortable that this is causing panic, since AFAICT all of the original code could be replicated in user WebGL. That said, this is a definite improvement. @bors-servo r+

**Comment by asajeffrey** [comment]
Ah, does bors not listen to review comments? @bors-servo r+

**Comment by jdm** [comment]
That's a good point. I should put together some testcases to experiment; filed https://github.com/servo/servo/issues/24244 to do that.

**Comment by jdm** [comment]
r? @asajeffrey

The addition of RGB to the framebuffer completion checks exposed some additional implementation holes which were easy enough to clean up while I was investigating them.

File: tests/wpt/webgl/meta/conformance/extensions/oes-texture-half-float.html.ini [hunk 1]
```diff
@@ -1,58 +1,24 @@
...
-    expected: FAIL
-
-  [WebGL test #93: gl.getParameter(gl.IMPLEMENTATION_COLOR_READ_FORMAT) should be null (of type object). Was 6408 (of type number).]
-    expected: FAIL
-
-  [WebGL test #96: getError expected: INVALID_OPERATION. Was NO_ERROR : IMPLEMENTATION_COLOR_READ_TYPE should fail for incomplete framebuffers.]
-    expected: FAIL
-
-  [WebGL test #82: getError expected: INVALID_OPERATION. Was NO_ERROR : IMPLEMENTATION_COLOR_READ_FORMAT should fail for incomplete framebuffers.]
-    expected: FAIL
```
Comment (asajeffrey) [inline tests/wpt/webgl/meta/conformance/extensions/oes-texture-half-float.html.ini]: Oooh!

**Review by asajeffrey — APPROVED** [review]
LGTM

**Comment by CYBAI** [comment]
```
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_true: getError expected: INVALID_OPERATION. Was INVALID_FRAMEBUFFER_OPERATION : After ReadPixels from missing attachment expected true got false", 
    "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1611:25\ntest@http://web-platform.test:8000/resources/testharness.js:544:30\nreportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:7\ntestFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:5\nglErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:5\nglErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10\ntestReadingFromMissingAttachment@http://web-platform.test:8000/_webgl/conformance/renderbuffers/framebuffer-object-attachment.html:647:9\n@http://web-platform.test:8000/_webgl/conformance/renderbuffers/framebuffer-object-attachment.html:430:5\n", 
    "subtest": "WebGL test #555: getError expected: INVALID_OPERATION. Was INVALID_FRAMEBUFFER_OPERATION : After ReadPixels from missing attachment", 
    "test": "/_webgl/conformance/renderbuffers/framebuffer-object-attachment.html", 
    "line": 186140, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/_webgl/conformance/glsl/bugs/complex-glsl-does-not-crash.html", 
    "line": 17647, 
    "action": "test_result", 
    "expected": "OK"
}
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Overall test", 
    "test": "/_webgl/conformance/glsl/misc/shader-with-non-reserved-words.html", 
    "line": 29943, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/_webgl/conformance/glsl/misc/shader-with-non-reserved-words.html", 
    "line": 32736, 
    "action": "test_result", 
    "expected": "OK"
}
```

**Comment by jdm** [comment]
The new commit (31c9d934fc3a093369f7957998a0302f93fb872d) is not strictly necessary to solve those particular test failures. It does address a hole in the current implementation, where changing the storage format of a framebuffer attachment does not cause the completion status to change.

**Comment by jdm** [comment]
Hold up - I'm getting weird results in the HL emulator that I need to investigate.

**Comment by jdm** [comment]
Verified that this works as expected in the emulator.

@nox could you review adaaadf?

File: components/script/dom/webglrenderbuffer.rs [hunk 2]
```diff
@@ -10,14 +10,16 @@ use crate::dom::bindings::codegen::Bindings::WebGLRenderbufferBinding;
...
+use crate::dom::bindings::root::{DomRoot, MutNullableDom};
+use crate::dom::webglframebuffer::WebGLFramebuffer;
 use crate::dom::webglobject::WebGLObject;
 use crate::dom::webglrenderingcontext::WebGLRenderingContext;
 use canvas_traits::webgl::{
     webgl_channel, GlType, WebGLCommand, WebGLError, WebGLRenderbufferId, WebGLResult,
 };
 use dom_struct::dom_struct;
 use std::cell::Cell;
+use std::default::Default;
```
Comment (nox) [inline components/script/dom/webglrenderbuffer.rs]: Nit: this is in prologue.

File: components/script/dom/webgltexture.rs [hunk 3]
```diff
@@ -10,15 +10,17 @@ use crate::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGL
...
 use crate::dom::webgl_validations::types::TexImageTarget;
+use crate::dom::webglframebuffer::WebGLFramebuffer;
 use crate::dom::webglobject::WebGLObject;
 use crate::dom::webglrenderingcontext::WebGLRenderingContext;
 use canvas_traits::webgl::{webgl_channel, TexDataType, TexFormat, WebGLResult, WebGLTextureId};
 use canvas_traits::webgl::{DOMToTextureCommand, WebGLCommand, WebGLError};
 use dom_struct::dom_struct;
 use std::cell::Cell;
 use std::cmp;
+use std::default::Default;
```
Comment (nox) [inline components/script/dom/webgltexture.rs]: Nit: this is in prologue.

**Review by nox — COMMENTED** [review]
r=me adaaadf with some nits.

**Pull request merged**